### PR TITLE
feat(datasource): implement custom datasource

### LINKED ISSUE #10622: Generic datasource
**What would you like Renovate to be able to do?**

Support the concept of a "generic" datasource, so that users can create an API matching its schema to massage unsupported datasources into something Renovate understands.

**Did you already have any implementation ideas?**

We take our existing "internal" datasource schema - i.e. the type definition for how datasources return results - and make that external/public in the form of a JSON schema potentially. Then users can e.g. scrape sources and publish them to a static website for Renovate to read.

Example:

Let's say we want to know the version of python supported by `pyenv`, and those are kept internal to their repo. So we create our own repo, called `pyenv-versions` and add a GitHub Action to run hourly and "scrape" the pyenv repo and generate a simple JSON file representing the versions, and then publish that to GitHub pages.

I can see two structures being supported:
1. One JSON file per package, so then you fetch `<registryUrl><lookupName>.json` and use the result directly
2. A single JSON file for all packages, so you fetch `<registryUrl>` and then use the result which exists inside the `lookupName` key in the JSON.

### PR DESCRIPTION
**Pull request by secustor**
## Changes
Implements a custom datasource which currently supports json http endpoints and transformations with jsonata

Feedback issue: https://github.com/renovatebot/renovate/discussions/23286

## Context
Supersedes https://github.com/renovatebot/renovate/pull/20753
Closes https://github.com/renovatebot/renovate/issues/10622

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

https://github.com/secustor/renovate-generic-datasource

### CODE REVIEW

**Comment by secustor** [comment]
The coverage seems to be reported falsely again. Locally the tests pass.

File: docs/usage/configuration-options.md [hunk 1]
```diff
@@ -633,6 +633,37 @@ When using with `npm`, we recommend you:
...
+    foo: {
+      defaultRegistryUrlTemplate: 'https://exmaple.foo.bar/v1/{{ packageName }}',
+    },
+  },
+}
+```
+
+### format
+
+Defines which format the API is returning. Available values: `json`
```
Comment (rarkins) [inline docs/usage/configuration-options.md:660]:
```suggestion
Defines which format the API is returning.
Only `json` is supported, but more are planned for future.
```

File: lib/config/options/index.ts [hunk 2]
```diff
@@ -923,6 +940,16 @@ const options: RenovateOptions[] = [
     cli: false,
     env: false,
   },
+  {
+    name: 'defaultRegistryUrlTemplate',
+    description:
+      'Template for generating a defaultRegistryUrl for custom datasource',
+    type: 'string',
+    default: '',
+    parent: 'customDatasources',
```
Comment (rarkins) [inline lib/config/options/index.ts:949]: Good! This way we can get it out there and experiment/iterate

**Review by rarkins — CHANGES_REQUESTED** [review]
I'm happy to get this merged and available as experimental as soon as we can

**Review by rarkins — APPROVED** [review]

**Pull request merged**